manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,956 @@
+				log.G(ctx).WithError(err).Error("node cannot be allocated")
+			}
+		}
+
+		// if any nodes successfully allocated, commit them
+		if len(allocatedNodes) > 0 {
+			if err := a.store.Batch(func(batch *store.Batch) error {
+				for _, node := range allocatedNodes {
+					writeTxDone := metrics.StartTimer(storeLockHeld.WithValues("node", "allocate", "write"))
+					if err := batch.Update(func(tx store.Tx) error {
+						currentNode := store.GetNode(tx, node.ID)
+						if currentNode == nil {
+							// if there is no node, then it must have been
+							// deleted. deallocate our changes.
+
+							// TODO(dperny): IMPORTANT! there is a race
+							// condition here which is very dangerous! calls to
+							// deallocate are not idempotent! this means that
+							// we can't be sure that we won't inadvertently
+							// free some state when we deallocate a second
+							// time! a classic double-free case!
+							log.G(ctx).WithField("node.id", node.ID).Debug("node was deleted after allocation but before committing")
+							a.network.DeallocateNode(node)
+							return nil
+						}
+
+						// the node may have changed in the meantime since we
+						// read it before allocation. but since we're the only
+						// one who changes the attachments, we can just plop
+						// our new attachments into that slice with no danger
+						currentNode.Attachments = node.Attachments
+
+						return store.UpdateNode(tx, currentNode)
+					}); err != nil {
+						log.G(ctx).WithError(err).WithField("node.id", node.ID).Error("error in committing allocated node")
+					}
+					writeTxDone()
+				}
+				return nil
+			}); err != nil {
+				log.G(ctx).WithError(err).Error("error in batch update of allocated nodes")
+			}
+		}
+	}
+
+	// finally, allocate all tasks.
+	if len(pendingTasks) > 0 {
+		// this might seem a bit nonobvious. basically, what we're doing is
+		// allocating services as needed by tasks. and we're going to optimize
+		// on allocating many tasks belonging to a service at the same time.
+		// So, in our View transaction, we're going to go through each pending
+		// task ID. if it's still ready for allocation, we're going to get its
+		// service and stash the service in a slice. then, we're going to stash
+		// the task in a slice in a map.
+
+		// services is a slice of services with tasks pending allocation
+		services := []*api.Service{}
+
+		// tasks is a map from service id to a slice of task objects belonging
+		// to that service
+		tasks := map[string][]*api.Task{}
+		storeDone := metrics.StartTimer(storeLockHeld.WithValues("service", "allocate", "read"))
+		a.store.View(func(tx store.ReadTx) {
+			for taskid := range pendingTasks {
+				t := store.GetTask(tx, taskid)
+				// if the task is not nil, is in state NEW, and is desired to
+				// be running, then it still needs allocation
+				if t != nil && t.Status.State == api.TaskStateNew && t.DesiredState == api.TaskStateRunning {
```

Comment:
What about tasks with a desired state of `READY`?

Reply:
good catch.